### PR TITLE
rustdoc: normalise type/field names

### LINKED ISSUE #100961: Rustdoc-Json: More inconsistant names
- `use`'s are refered to as `Import`'s
- `FnDecl` uses short `Fn`, but `Function` and `FunctionPointer` don't abbreviate
- `OpaqueTy` uses short `Ty`, but full `type` uses everywhere else
- `Header` not called `FunctionHeader`
- `Header` has fields `async_`, `const_` and `unsafe_`, but should be `is_async` (etc), like rest of types.

Somewhat of a followup to #94889

@rustbot modify labels: +T-rustdoc +A-rustdoc-json

### PR DESCRIPTION
**Pull request by its-the-shrimp**
Updates #100961 

- `Import` -> `Use`, to better reflect the terminology of Rust & its syntax
- `TypeBinding` -> `AssocItemConstraint`, to sync up with `clean`
- `FnDecl` -> `FunctionSignature`, because that's what it is
- `Header` -> `FunctionHeader`, because `Header` is a very word that's very heavily loaded with different meanings
- `ItemEnum::AssocType`: `default` -> `type`, because those items appear in `impl` blocks as well, where they're _not_ the "default"
- `ItemEnum::AssocConst`: `default` -> `value`, see the previous point
- `ForeignType` -> `ExternType`, because "foreign" is not the right word there
- boolean fields' names made to consistently be a phrase that can be a yes/no answer, e.g. `async` -> `is_async`

The docs of `ItemEnum::AssocType::type_` & of `ItemEnum::AssocConst::value` are also updated to be up to date with the clarification of the name of the fields

### CODE REVIEW

**Comment by camelid** [comment]
Hmm, some of this feels like unnecessary churn - particularly the changes to the boolean field names. For example, I don't think `is_async` is any clearer than `async`. I also don't see the point of renaming `FnDecl` since that's what it's called in rustc AFAICT. Curious what others think though.

**Comment by workingjubilee** [comment]
Hm. I agree `is_async` is not clearer than `async` but I think it **is** clearer than `async_`, which just looks like a typo.

**Comment by obi1kenobi** [comment]
I love the effort to use clearer and more consistent terminology, and document rustdoc JSON types more thoroughly! Thanks to everyone helping this along 🙏

If possible, I'd just ask that you double-check that the name changes didn't negatively affect the size of generated rustdoc JSON files, e.g. by making an enum discriminant string substantially longer than before.

`cargo-semver-checks` is used on some crates with as many as ~250k items (e.g. `aws-sdk-ec2` or `windows`), and currently the rule of thumb for rustdoc JSON is ~1KB/item on average. When the JSON file is 250MB and we need to compare two of them to each other on CI-grade hardware, IO size and parsing time are a significant factor in overall runtime. I wouldn't ask to squeeze out every last byte — I'd just hope to avoid things like "added 10 extra bytes apiece in something that'll appear 250k times in the file" because that would be a measurable slowdown. To be clear, I don't think that's happened here, but I'd love to be extra sure.

**Comment by its-the-shrimp** [comment]
The reason for renaming `FnDecl`, and the whole PR in general, is to be more friendly to outside users, that might not be in the know about rustc conventions, a function declaration may be understood as `fn printf(fmt: &str, ...);`,even though what the `FnDecl` is there is actually just `(fmt: &str, ...)`, which is universally understood to be a function signature

**Comment by its-the-shrimp** [comment]
regarding boolean fields, it's fine if we end up deciding to keep their names as they are, however, another reason for this PR was to close the question of naming in rustdoc-json (hence the "Fixes" annotation at the start), thus, I believe whatever renames we happen to decide on in this PR will serve as a precedent for naming conventions for rustdoc-json in the future (I believe @aDotInTheVoid wanted to document those conventions)

File: src/librustdoc/clean/mod.rs [hunk 1]
```diff
@@ -1143,7 +1143,7 @@ fn clean_fn_decl_with_args<'tcx>(
     decl: &hir::FnDecl<'tcx>,
     header: Option<&hir::FnHeader>,
     args: Arguments,
-) -> FnDecl {
+) -> FunctionSignature {
```
[the remark this answers is not in the document]
Comment (fbstj) [reply src/librustdoc/clean/mod.rs:1146]: > The reason for renaming FnDecl, and the whole PR in general, is to be more friendly to outside users, that might not be in the know about rustc conventions, a function declaration may be understood as fn printf(fmt: &str, ...);,even though what the FnDecl is there is actually just (fmt: &str, ...), which is universally understood to be a function signature

am I right in thinking that anyone using this output will know enough rust to relate the `fn` keyword to functions, and probably the `Fn*` traits.

[the remark this answers is not in the document]
Comment (its-the-shrimp) [reply src/librustdoc/clean/mod.rs:1146]: well, `fn` is a fundamental keyword in Rust, so `fn` <-> `function` is a trivial expectation from the user, the change from `Fn` to `Function` there is purely for consistency with the existing `Function` & `FunctionPointer` structures, I don't believe we should be swayed by rustc's naming in rustdoc-json, since rustc is definitely not intended to be public API, unlike rustdoc-json

[the remark this answers is not in the document]
Comment (its-the-shrimp) [reply src/librustdoc/clean/mod.rs:1146]: And the change from `Decl` to `Signature` is explained in my comment a bit below

**Review by aDotInTheVoid — CHANGES_REQUESTED** [review]
While there's value for making all the JSON changes in one PR (we batch many small breaking changes into 1 large one), it's not a good idea for the `clean` types. If you want to make changes to them, they can be done in later, smaller PR's that just do one thing (so are easier to review).

File: src/rustdoc-json-types/lib.rs [hunk 2]
```diff
@@ -258,19 +258,19 @@ pub struct Constant {
 ///              ^^^^^^^^^^  ^^^^^^^^^^^^^^^
 /// ```
 #[derive(Clone, Debug, PartialEq, Eq, Hash, Serialize, Deserialize)]
-pub struct TypeBinding {
+pub struct AssocItemConstraint {
```
Comment (aDotInTheVoid) [inline src/rustdoc-json-types/lib.rs:261]: Why this change?

Comment (its-the-shrimp) [reply src/rustdoc-json-types/lib.rs:261]: a "type binding" is a confusing name, and frankly outdated, since nowadays it can also constrain a constant, and it only ever operates on associated items, by constraining them

**Comment by its-the-shrimp** [comment]
Should I revert the changes in jsondoclint as well?

**Comment by aDotInTheVoid** [comment]
> Should I revert the changes in jsondoclint as well?

No, jsondoclint uses the types from rustdoc_json_types, so it will need to be changed.

**Review by obi1kenobi — COMMENTED** [review]
While I am generally in favor of the renaming of what the fields are called in _Rust_, I am strongly opposed to the new longer field names _in JSON_.

I'd like to ask that we consider not introducing the `is_` prefixes into the rustdoc JSON itself, to avoid bloating the file and causing perf regressions without any human-visible improvement.

The largest crate that uses `cargo-semver-checks` produces a 475MB rustdoc JSON file, and contains ~100k functions and ~400k trait and inherent impls. Deserializing that file is a measurable bottleneck in `cargo-semver-checks` — we deserialize two such files per run, and spend about 4-5s wall-clock on it.

The proposed changes in the JSON field names in this PR will increase the file size by about 3-4MB (~1%). There's no real offsetting benefit for this regression, since the JSON is primarily meant to be machine-readable — it isn't even pretty-printed by default. There's very little advantage to having more descriptive field names repeat thousands of times.

Let's please keep the JSON field names short?

**Comment by its-the-shrimp** [comment]
That's valid, I was also thinking about that, and I think there's a way to avoid the tug'o'war between the JSON & the Rust repr, which is having certain fields not serialised if they contain the "default" value whatever it happens to be, e.g. `false` for booleans, I think this will bring compression benefits that'll far outweigh any size increases brought by long field names

E.g.

```rust
    #[serde(default)]
    #[serde(skip_serializing_if = "is_default")] // Function to be defined
    pub is_async: bool,
```

**Comment by obi1kenobi** [comment]
I'd love to have both default/omitted fields and short field names as much as possible!

475MB of rustdoc JSON is _a lot_, and the scary thing is that a year ago that crate "only" produced ~250MB of rustdoc JSON. I'm not sure how much of that is due to format changes vs growth in the crate, but it's not a trend we can sustain for long either way. Short field names are a cheap way to buy a few percent, and especially leaving JSON field names unchanged by only adding the `is_` prefix in Rust not JSON feels like an easy win.

**Comment by its-the-shrimp** [comment]
@rustbot ready
I've also renamed the `can_unwind` field in `Abi` back to `unwind` to not have the name bring any assumptions about the technical details of the ABI by the word "can", and I've reverted all the changes in the HTML output of rustdoc

**Comment by its-the-shrimp** [comment]
Looking at it now, I'm pretty sure that omitted fields could bring size improvements of much more than the 1 percent regression brought by the `is_` prefixes, there are a lot of fields carrying booleans, options, integers which employ 0 as a special value, arrays which could be omitted completely if empty, etc., then we could also omit the ABI if it's the Rust ABI, the possibilities are boundless

**Comment by obi1kenobi** [comment]
> Looking at it now, I'm pretty sure that omitted fields could bring size improvements of much more than the 1 percent regression brought by the `is_` prefixes

I agree, and I'm in favor of that. But I don't think it's an either-or situation, and I don't think adding the `is_` prefixes like this PR is doing is making rustdoc JSON any better off. I think it's better to not have them, regardless of whether we end up omitting fields later.

**Comment by its-the-shrimp** [comment]
@aDotInTheVoid what's your opinion on the changes to the boolean field names?

**Comment by aDotInTheVoid** [comment]
I think that if changing the field names to be longer causes a perf regression, that's a problem with the overall serialization/de-serialization strategy, not the type design. I'm strongly opposed to changing field names or using serde defaults to try to fix perf. I want to review this PR solely based on how good the names are.

I've opened a [zulip thread](https://rust-lang.zulipchat.com/#narrow/stream/266220-t-rustdoc/topic/rustdoc-json.3A.20compressed.20output) to discuss potentially compressing or using alternative serialization formats for perf-sensitive users. Please don't have any further discussion around output sizes in this PR.

File: src/rustdoc-json-types/lib.rs [hunk 3]
```diff
@@ -697,11 +705,11 @@ pub enum Abi {
 #[derive(Clone, Debug, PartialEq, Eq, Hash, Serialize, Deserialize)]
 pub struct Function {
     /// Information about the function signature, or declaration.
-    pub decl: FnDecl,
+    pub decl: FunctionSignature,
```
Comment (aDotInTheVoid) [inline src/rustdoc-json-types/lib.rs:708]: Given this, we should probably rename the field to `signature` or `sig`

File: src/rustdoc-json-types/lib.rs [hunk 4]
```diff
@@ -1036,7 +1044,7 @@ pub struct Path {
 #[derive(Clone, Debug, PartialEq, Eq, Hash, Serialize, Deserialize)]
 pub struct FunctionPointer {
     /// The signature of the function.
-    pub decl: FnDecl,
+    pub decl: FunctionSignature,
```
Comment (aDotInTheVoid) [inline src/rustdoc-json-types/lib.rs:1047]: Ditto

**Review by aDotInTheVoid — COMMENTED** [review]
This looks really close, thanks for working on it, and sorry for the review delay.

r=me with field name renamed

File: src/llvm-project
Comment (aDotInTheVoid) [inline src/llvm-project:1]: You managed to accidently update the llvm-project submodule somehow. This needs to be reverted before this can land.

**Comment by aDotInTheVoid** [comment]
@bors r+


Thanks so much for working on this!
